L1TriggerScouting/Phase2/plugins/ScPhase2RecMesonAll.cc

```diff
@@ -0,0 +1,275 @@
+      maxDeltaRDaus_(iConfig.getParameter<double>("maxDeltaRDaus")),
+      maxDeltaZ_(iConfig.getParameter<double>("maxDeltaZ")),
+      minPtDau_(iConfig.getParameter<double>("minPtDau")),
+      maxZIsolation_(iConfig.getParameter<double>("maxZIsolation"))
+  {
+  if (doStruct_) {
+    //PUPPI input being given here
+    structToken_ = consumes<OrbitCollection<l1Scouting::Puppi>>(iConfig.getParameter<edm::InputTag>("src"));
+
+    std::vector<edm::ParameterSet> mesonPsets =
+        iConfig.getParameter<std::vector<edm::ParameterSet>>("mesonTypes");
+
+    for (const auto& pset : mesonPsets) {
+      mesonTypeStruct mt;
+      mt.name = pset.getParameter<std::string>("name");
+      mt.minMesonMass = pset.getParameter<double>("minMesonMass");
+      mt.maxMesonMass = pset.getParameter<double>("maxMesonMass");
+      mt.dauMass1 = pset.getParameter<double>("dauMass1");
+      mt.dauMass2 = pset.getParameter<double>("dauMass2");
+      mesonTypes_.push_back(mt);
+
+      // Register output collections with instance labels
+      produces<OrbitCollection<l1Scouting::RecMeson<2>>>(mt.name);
+    }
+  }
+}
+
+ScPhase2RecMesonAll::~ScPhase2RecMesonAll() {};
+
+void ScPhase2RecMesonAll::beginStream(edm::StreamID) {}
+
+void ScPhase2RecMesonAll::produce(edm::Event &iEvent, const edm::EventSetup &iSetup) {
+  if (doStruct_) {
+    edm::Handle<OrbitCollection<l1Scouting::Puppi>> src;
+    iEvent.getByToken(structToken_, src);
+
+    runObj(*src, iEvent, "");
+  }
+}
+
+void ScPhase2RecMesonAll::endStream() {}
+
+template <typename T>
+void ScPhase2RecMesonAll::runObj(const OrbitCollection<T> &src,
+                                    edm::Event &iEvent,
+                                    const std::string &label) {
+  // l1ScoutingRun3::BxOffsetsFillter bxOffsetsFiller;
+  // bxOffsetsFiller.start();
+  auto ret = std::make_unique<std::vector<unsigned>>();
+
+  ROOT::RVec<unsigned int> ix;
+
+  std::vector<std::vector<std::vector<l1Scouting::RecMeson<2>>>> all_mesonVec;
+  std::vector<unsigned int> all_ntotRecMeson(mesonTypes_.size(), 0);;
+  int nbx = 0;
+
+  for (unsigned int bx = 0; bx <= OrbitCollection<T>::NBX; ++bx) {
+    nbx++;
+    std::vector<std::vector<l1Scouting::RecMeson<2>>> all_mesonVec_thisBx(mesonTypes_.size());
+
+    auto range = src.bxIterator(bx);
+    const T *cands = &range.front();
+    auto size = range.size();
+
+    ix.clear();
+    for (unsigned int i = 0; i < size; ++i) {  //make list of all hadrons
+      if ((std::abs(cands[i].pdgId()) == 211 or std::abs(cands[i].pdgId()) == 11)) {
+        if (cands[i].pt() >= minPtDau_)
+          ix.push_back(i);
+      }
+    }
+    unsigned int ndaus = ix.size();
+
+    for (unsigned int i1 = 0; i1 < ndaus; ++i1) {
+      for (unsigned int i2 = i1 + 1; i2 < ndaus; ++i2) {
+        if (!(cands[ix[i1]].charge() * cands[ix[i2]].charge() < 0))
+          continue;
+
+        float drQ = deltar(cands[ix[i1]].eta(), cands[ix[i2]].eta(), cands[ix[i1]].phi(), cands[ix[i2]].phi());
```

Comment:
```suggestion
        float drQ = reco::deltaR(cands[ix[i1]], cands[ix[i2]]);
```
(from `#include "DataFormats/Math/interface/deltaR.h"`)